[change] Merging with wrong format now raises Validation error #351

stktyagi opened the pull request:
An attempt to merge config with wrong format will now raise a validation error instead of failing.

Fixes #351

## Checklist

- [x] I have read the [OpenWISP Contributing Guidelines](http://openwisp.io/docs/developer/contributing.html).
- [x] I have manually tested the changes proposed in this pull request.
- [x] I have written new test cases for new code and/or updated existing tests for changes to existing code.
- [ ] I have updated the documentation.

## Reference to Existing Issue

Closes #351 

## Description of Changes

Added conditionals to check if two template types can be merged and if not raise validation error instead of failing.

stktyagi committed:
[Fix] Merging with wrong format now raises Validation error #351

An attempt to merge config with wrong format will now raise a validation error instead of failing.

Fixes #351

nemesifier commented on netjsonconfig/utils.py:
```diff
@@ -3,6 +3,12 @@
 from copy import deepcopy
 
 
+class ValidationError(Exception):
```
Duplicate of `netjsonconfig.exceptions.ValidationError`.

stktyagi committed:
Merge branch 'master' into master

stktyagi committed:
[Refactor] Change aggressive validation in merge_config #351

It only prevents the original AttributeError crash and correctly lets the .validate() method handle all validation of the final merged configuration.

Fixes #351

stktyagi committed:
Merge branch 'master' into master

nemesifier reviewed (changes requested):
@stktyagi the description says:

An attempt to merge config with wrong format will now raise a validation error instead of failing.

However, I don't see this in this implementation, nor in the test. No validation error is raised, it looks as if the error is silently ignored, which is not what I was expecting.

Why this inconsistency? Is this intentional?

stktyagi committed:
Merge branch 'master' into master

stktyagi committed:
[fix] Riase ValidationError on incompatible types #351

Merge_config now raises ValidationError when templates have incompatible types for the same key (e.g. list vs dict), instead of silently overwriting.
Test updated to assert ValidationError.

Fixes #351

stktyagi committed:
[fix] Scope type check to container-type conflicts onl #351

Narrowed the incompatible type check in merge_config to only
raise ValidationError when a dict or list conflicts with a
different type. Scalar overrides (e.g. str to int) are still
allowed as before.

Fixes #351

stktyagi committed:
Merge branch 'master' into master

nemesifier commented on netjsonconfig/utils.py:
```diff
@@ -19,14 +23,26 @@ def merge_config(template, config, list_identifiers=None):
+            and type(value) is not type(existing)
+        ):
+            raise ValidationError(
+                JsonSchemaError(
```
This constructs a `JsonSchemaError` with only a message. That gives the right exception type, but `netjsonconfig.exceptions.ValidationError.__str__` later assumes the wrapped jsonschema error has iterable `validator_value` and `context`. For this path, `str(exc)` raises `TypeError: 'Unset' object is not iterable`, so an uncaught error is shown as `<exception str() failed>` instead of the validation message. Please either make the wrapper tolerate message only errors, or build the wrapped error with the fields that `__str__` expects. The test should also assert that the raised error can be stringified.

nemesifier commented on netjsonconfig/utils.py:
```diff
@@ -19,14 +23,26 @@ def merge_config(template, config, list_identifiers=None):
+            result[key] = merge_config(existing, value, list_identifiers)
+        elif isinstance(value, list) and isinstance(existing, list):
+            result[key] = merge_list(existing, value, list_identifiers)
+        elif (
```
This guard applies at every recursive merge level, not only to the top level package shape conflict from issue 351. It now rejects cases where a matched custom UCI block overrides a list option with a scalar option, even though the merge contract says simple values in the config overwrite values in the template. For example, a template dnsmasq block with `server: ["1.1.1.1"]` and device config with `server: "8.8.8.8"` used to merge, but now raises because `existing` is a list and `value` is a string. Please scope the validation to the package shape conflict or other cases that would otherwise crash, unless this broader behavior change is intentional and documented.

nemesifier reviewed (changes requested):
The direct reproduction from issue #351 now raises the project ValidationError, but I found two remaining problems that should be fixed before merge. See my comments below.

stktyagi committed:
Merge branch 'master' into master

stktyagi committed:
[fix] Fix uncaught exception on incompatible merge types

Make NetJsonConfigException tolerant of message-only jsonschema errors to prevent TypeError during stringification. Scope type mismatch validation in merge_config strictly to complex shape conflicts (dict vs list).

nemesifier committed:
[chores] Minor improvements
